explainability-core/src/main/java/org/kie/trustyai/metrics/drift/fouriermmd/FourierMMD.java

```diff
@@ -0,0 +1,504 @@
+                for (int c = 0; c < numColumns; c++) {
+                    xWindowed[r - idx][c] = xIn.getRow(r).get(c).asNumber();
+                }
+            }
+
+            // x1 = x1 / self.learned_params["scale"].repeat(x1.shape[0], 0)
+
+            for (int r = 0; r < n_window; r++) {
+                for (int c = 0; c < numColumns; c++) {
+                    xWindowed[r][c] /= scaleArray[c];
+                }
+            }
+
+            // r_cos = np.cos(np.matmul(x1, wave_num) + bias.repeat(x1.shape[0], 0))
+
+            final Array2DRowRealMatrix xWindowedMatrix = new Array2DRowRealMatrix(xWindowed);
+            final Array2DRowRealMatrix product2 = xWindowedMatrix.multiply(waveNumMatrix);
+
+            final double[][] rCos2 = new double[n_window][n_mode];
+
+            for (int r = 0; r < n_window; r++) {
+                for (int c = 0; c < n_mode; c++) {
+                    final double entry = product2.getEntry(r, c);
+                    final double newEntry = entry + bias[0][c];
+                    rCos2[r][c] = Math.cos(newEntry);
+                }
+            }
+
+            // a_comp = r_cos.mean(0) * np.sqrt(2 / self.n_mode)
+
+            final double[] aComp = new double[n_mode];
+            final double multiplier2 = Math.sqrt(2.0 / n_mode);
+            for (int c = 0; c < n_mode; c++) {
+                double sum = 0.0;
+                for (int r = 0; r < n_window; r++) {
+                    sum += rCos2[r][c];
+                }
+
+                aComp[c] = (sum / n_window) * multiplier2;
+            }
+
+            // mmd = ((a_ref - a_comp) ** 2).sum()
+
+            double mmd = 0.0;
+            for (int c = 0; c < n_mode; c++) {
+                final double dist = aRef[c] - aComp[c];
+
+                final double term = dist * dist;
+                mmd += term;
+            }
+
+            // sample_mmd += [mmd]
+
+            sampleMMD[i] = mmd;
+        }
+
+        // self.learned_params["mean_mmd"] = np.nanmean(np.array(sample_mmd))
+
+        fitStats.mean_mmd = 0.0;
+        final List<Double> sampleMMD2 = new ArrayList<Double>(n_test);
+        for (int i = 0; i < n_test; i++) {
+            if (!Double.isNaN(sampleMMD[i])) {
+                sampleMMD2.add(sampleMMD[i]);
+            }
+        }
+
+        if (sampleMMD2.size() == 0) {
+            throw new IllegalArgumentException("sampleMMD2 length is zero");
+        }
+
+        final double[] sampleMMD2NoNaN = new double[sampleMMD2.size()];
+        for (int i = 0; i < sampleMMD2.size(); i++) {
+            sampleMMD2NoNaN[i] = sampleMMD2.get(i);
+        }
+
+        fitStats.mean_mmd = mean.evaluate(sampleMMD2NoNaN);
+
+        // self.learned_params["std_mmd"] = np.nanstd(np.array(sample_mmd))
+
+        fitStats.std_mmd = std.evaluate(sampleMMD2NoNaN);
+
+        // return self.learned_params
+
+        return fitStats;
+    }
+
+    public DriftResult calculate(Dataframe data) {
```

Comment:
Same here

Reply:
I factored a bunch of code out of it.  If you want I can break it up more....